glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/DeletionPingUploadWorker.kt

```diff
@@ -0,0 +1,192 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.telemetry.glean.scheduler
+
+import android.content.Context
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import androidx.work.Constraints
+import androidx.work.ExistingWorkPolicy
+import androidx.work.NetworkType
+import androidx.work.OneTimeWorkRequest
+import androidx.work.OneTimeWorkRequestBuilder
+import androidx.work.WorkManager
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import mozilla.telemetry.glean.Glean
+import java.io.BufferedReader
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.FileReader
+import java.io.IOException
+
+/**
+ * This class is the worker class used by [WorkManager] to handle uploading the ping to the server.
+ * @suppress This is internal only, don't show it in the docs.
+ */
+class DeletionPingUploadWorker(context: Context, params: WorkerParameters) : Worker(context, params) {
+    companion object {
+        internal const val PING_WORKER_TAG = "mozac_service_glean_deletion_ping_upload_worker"
+
+        // Since ping file names are UUIDs, this matches UUIDs for filtering purposes
+        private const val FILE_PATTERN = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        private const val LOG_TAG = "glean/DeletionPingUploadWorker"
+        // NOTE: The `PINGS_DIR` must be kept in sync with the one in the Rust implementation.
+        internal const val PINGS_DIR = "deletion_request"
+        // A lock to prevent simultaneous writes in the ping queue directory.
+        // In particular, there are issues if the pings are cleared (as part of
+        // disabling telemetry), while the ping uploader is trying to upload queued pings.
+        // Therefore, this lock is held both when uploading pings and when calling
+        // into the Rust code that might clear queued pings (set_upload_enabled).
+        internal val pingQueueLock = Any()
+
+        /**
+         * Build the constraints around which the worker can be run, such as whether network
+         * connectivity is required.
+         *
+         * @return [Constraints] object containing the required work constraints
+         */
+        @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+        fun buildConstraints(): Constraints = Constraints.Builder()
```

Comment:
Since this is likely to be shared across the two workers, what if we just used the function from the main upload worker class?

Reply:
Will refactor this to share more of the code

Reply:
This can also happen as a follow-up, depending on your will to iterate on this PR :)

Reply:
Follow-up it is! filing bug now